### ISSUE TITLE
`clippy::bytes_nth` automatic fix causes compile errors

### ISSUE THREAD

**Issue by alexkeizer** [rust-lang/rust-clippy#10151]
### Summary

The automatic fix for the `clippy::bytes_nth` lint replaces 
```rust
bar.bytes().nth(1).unwrap()
```
with 
```
bar.as_bytes().get(1).unwrap()
```

Failing to notice that the former is a `u8` while the latter is a `&u8`.
It should automatically add the dereference as well, i.e., 
```
*bar.as_bytes().get(1).unwrap()
```

EDIT: Since `as_bytes` gives a slice, the `.get(i).unwrap()` can of course be just `[i]`
```
bar.as_bytes()[i]
```
Which seems like a clear improvement over the original.

In the more general case (i.e., without `unwrap`), though, should
```
bar.bytes().nth(1)
```
be replaced by the following?
```
bar.bytes().get(1).copied()
```

### Reproducer

I tried to run `cargo clippy --fix` on the following MWE:

```rust
fn foo(x : u8) -> u8 {
    x
}

fn main() {
    let bar = "baz";
    foo(bar.bytes().nth(1).unwrap());
}
```

And was met with the following output:
```
warning: failed to automatically apply fixes suggested by rustc to crate `clippy_big_mwe`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see 
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
 --> src/main.rs:8:9
  |
8 |     foo(bar.as_bytes().get(1).unwrap());
  |     --- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u8`, found `&u8`
  |     |
  |     arguments to this function are incorrect
  |
note: function defined here
 --> src/main.rs:2:4
  |
2 | fn foo(x : u8) -> u8 {
  |    ^^^ ------
help: consider dereferencing the borrow
  |
8 |     foo(*bar.as_bytes().get(1).unwrap());
  |         +

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: called `.bytes().nth()` on a `str`
 --> src/main.rs:8:9
  |
8 |     foo(bar.bytes().nth(1).unwrap());
  |         ^^^^^^^^^^^^^^^^^^ help: try: `bar.as_bytes().get(1)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#bytes_nth
  = note: `#[warn(clippy::bytes_nth)]` on by default

warning: `clippy-big-mwe` (bin "clippy-big-mwe" test) generated 1 warning (run `cargo clippy --fix --bin "clippy-big-mwe" --tests` to apply 1 suggestion)
warning: `clippy-big-mwe` (bin "clippy-big-mwe") generated 1 warning (1 duplicate)
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s

```

### Version

```text
$ rustc --version --verbose`
rustc 1.68.0-nightly (d6f99e535 2023-01-02)
binary: rustc
commit-hash: d6f99e535a301a421dfee52a7c25bb4bdf420344
commit-date: 2023-01-02
host: x86_64-unknown-linux-gnu
release: 1.68.0-nightly
LLVM version: 15.0.6

$ cargo clippy --version
clippy 0.1.68 (d6f99e5 2023-01-02)
```

**Comment by alexkeizer** [comment]
@sebschmi @TaKO8Ki I saw you were the ones involved in adding the `bytes_nth` lint, could you comment on this issue?

**Comment by nindalf** [comment]
Discussed this with @llogiq. The right behaviour here is probably:

- Replace `bytes().nth(n).unwrap()` with `as_bytes()[n]`
- Replace other `bytes().nth(n)` with `as_bytes().get(n).copied()`
